### PR TITLE
Feras/backend/get student data

### PR DESCRIPTION
**Pull request by zerfuth**
Endpoint for retrieving a student's major, minors, completed courses, missing core classes offered next year and desired tech comps offered next year.

### CODE REVIEW

File: backend/test/users_test.js [hunk 1]
```diff
@@ -623,40 +623,41 @@ describe("Tests add admin user", function() {
...
+//         major: [],
+//         minor: [],
+//         courses: [{ course_code: "ECSE 428", semester: "winter" }]
+//       };
+//       if (JSON.stringify(res) == JSON.stringify(searchingFor)) {
+//         found = true;
+//       }
+//       assert(true, found);
+//     });
+//   });
```
Comment (KevenLYF) [inline backend/test/users_test.js]: are there any operations for removing the test data inserted ?

File: backend/logic/users/users.js [hunk 2]
```diff
@@ -170,9 +194,9 @@ var getCompletedCourses = async studentID => {
...
   let error = false;
-  if (!format.verifyStudentId(studentID)) {
-    error = "invalid format id";
-  }
+  // if (!format.verifyStudentId(studentID)) {
+  //   error = "invalid format id";
+  // }
 
   if (!error == false) {
     console.error(error);
```
Comment (Ty-Won) [inline backend/logic/users/users.js]: With the way this is written now, the if statement will never run

File: backend/logic/users/users.js [hunk 3]
```diff
@@ -181,20 +205,117 @@ var getStudentData = async studentID => {
...
+      AND (semester = ? OR semester = ?) 
+      AND (course_code 
+          IN (SELECT course_code 
+              FROM curriculum_tech_comps 
+              WHERE curriculum_name = ?)) 
+      AND (course_code
+          IN (SELECT course_code 
+              FROM student_desired_courses 
+              WHERE student_id = ?));`,
+      [studentID, fallSem, winterSem, curriculumName, studentID]
```
Comment (Ty-Won) [inline backend/logic/users/users.js]: explain the query to me to help me understand it

File: backend/test/users_test.js (hunk 1 above)
Comment (zerfuth) [reply backend/test/users_test.js]: No this test is outdated hence why I commented it out. Whoever takes the task for testing this endpoint will have to rewrite most of it.

File: backend/logic/users/users.js (hunk 3 above)
Comment (zerfuth) [reply backend/logic/users/users.js]: Select all courses that are not in the set of completed courses AND that are in the correct semester (offered next year) AND that are in the set of valid techcomps for the major AND that are in the set of desired courses to take

File: backend/logic/users/users.js [hunk 4]
```diff
@@ -181,20 +205,117 @@ var getStudentData = async studentID => {
...
+      `SELECT curriculum_name FROM student_minors WHERE student_id = ?;`,
+      [studentID]
+    );
+
+    let currYear = new Date().getFullYear();
+    let currMonth = new Date().getMonth();
+    let fallSem, winterSem;
+    let curriculumName = major[0].curriculum_name;
+
+    if (currMonth < 3) {
```
Comment (si-x-li) [inline backend/logic/users/users.js]: What about Summer?

File: backend/logic/users/users.js (hunk 3 above)
Comment (si-x-li) [reply backend/logic/users/users.js]: Is there a way to reduce the number of tables being queried? Or can you pull data from the tables you need and put them in a hashtable? As more data gets inserted in the database, these queries are going to become way too slow

Comment (zerfuth) [reply backend/logic/users/users.js]: Reduced it by one. I know it's still not optimal but I suggest we just make a low priority task for further optimizing if that's fine given there are more urgent tasks for now.

File: backend/logic/users/users.js (hunk 4 above)
Comment (zerfuth) [reply backend/logic/users/users.js]: Current webscraping doesn't include summer courses so it should be fine but I agree that it should be supported. I suggest adding a low priority task for refactoring this code to include summer terms and minors later on :)

**Comment by Ty-Won** [comment]
@zerfuth resolved conflicts, please confirm these changes I've made are as desired. If so merge

**Comment by zerfuth** [comment]
seems good to me

File: backend/logic/users/users.js [hunk 5]
```diff
@@ -194,9 +194,9 @@ var getCompletedCourses = async studentID => {
...
 var getStudentData = async studentID => {
   let error = false;
-  // if (!format.verifyStudentId(studentID)) {
-  //   error = "invalid format id";
-  // }
+  if (!format.verifyStudentId(studentID)) {
+    error = "invalid format id";
+  }
 
   if (!error == false) {
```
Comment (KevenLYF) [inline backend/logic/users/users.js]: You can simply do 
```
let error;
...
if (error) {
// handler
}
```

File: backend/logic/users/users.js [hunk 6]
```diff
@@ -181,20 +181,111 @@ var getStudentData = async studentID => {
...
+              WHERE curriculum_name = ?)) 
+      AND (SELECT course_code 
+          FROM student_desired_courses 
+          WHERE (student_id = ?, desired = TRUE));`,
+      [studentID, fallSem, winterSem, curriculumName, studentID]);
+    
+    // For each incomplete course, get prereqs and coreqs
+    // for(course c in courses){
+    //   c[prereqs] = [];
+    //   c[coreqs] = [];
```
Comment (KevenLYF) [inline backend/logic/users/users.js]: you can remove the comments here

File: backend/logic/users/users.js (hunk 5 above)
Comment (zerfuth) [reply backend/logic/users/users.js]: true true

**Pull request merged**